home-application/src/main/java/com/softserveinc/ita/homeproject/application/mapper/config/CustomHomeMappingConfig.java

```diff
@@ -0,0 +1,7 @@
+package com.softserveinc.ita.homeproject.application.mapper.config;
+
+import org.modelmapper.ModelMapper;
+
+public interface CustomHomeMappingConfig {
```

Comment:
I think this interface could define a contract much stricter and be smarter.

What if we make it as the following:
```java
package com.softserveinc.ita.homeproject.application.mapper.config;

import org.modelmapper.TypeMap;
import org.springframework.core.GenericTypeResolver;

public interface HomeMappingConfig<S, D> {

    void addMappings(TypeMap<S, D> typeMap);

    @SuppressWarnings("unchecked cast")
    default Class<S> getSourceType() {
        return (Class<S>) getResolvedClasses()[0];
    }

    @SuppressWarnings("unchecked cast")
    default Class<D> getDestinationType() {
        return (Class<D>) getResolvedClasses()[1];
    }

    private Class<?>[] getResolvedClasses() {
        Class<?>[] typeArguments = GenericTypeResolver.resolveTypeArguments(this.getClass(), HomeMappingConfig.class);
        final var expectedTypeArgumentsSize = 2;
        if (typeArguments == null || typeArguments.length != expectedTypeArgumentsSize) {
            throw new IllegalStateException("Something went wrong with resolving types. Try to provide it explicitly.");
        }
        return typeArguments;
    }
}

```

Then we can use it in our mapper:

```java
package com.softserveinc.ita.homeproject.application.mapper;

import javax.annotation.PostConstruct;
import java.util.Set;

import com.softserveinc.ita.homeproject.application.mapper.config.HomeMappingConfig;
import lombok.RequiredArgsConstructor;
import org.modelmapper.ModelMapper;
import org.modelmapper.TypeMap;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class HomeMapper {

    private ModelMapper modelMapper;

    @SuppressWarnings("rawtypes")
    private final Set<HomeMappingConfig> homeMappingConfigs;


    @PostConstruct
    @SuppressWarnings({"unchecked cast", "rawtypes"})
    public void init() {
        modelMapper = new ModelMapper();
        for (HomeMappingConfig homeMappingConfig : homeMappingConfigs) {
            TypeMap typeMap = modelMapper.typeMap(homeMappingConfig.getSourceType(), homeMappingConfig.getDestinationType());
            homeMappingConfig.addMappings(typeMap);
        }
    }

    public <D> D convert(Object source, Class<D> destination) {
        return modelMapper.map(source, destination);
    }
}

```

And it gives us an opportunity to configure:

```java
package com.softserveinc.ita.homeproject.application.mapper.config;

import com.softserveinc.ita.homeproject.homedata.entity.User;
import com.softserveinc.ita.homeproject.homeservice.dto.UserDto;
import org.modelmapper.TypeMap;
import org.springframework.stereotype.Component;

@Component
public class UserToUserDtoMappingConfig implements HomeMappingConfig<User, UserDto> {

    @Override
    public void addMappings(TypeMap<User, UserDto> typeMap) {
        typeMap.addMapping(User::getContacts, UserDto::setContacts);
    }
}

```

Reply:
Thanks. You asked me to do it but I didn't figure out how to do it. Done.